### ISSUE TITLE
write_rubias bug (maybe just me)

### ISSUE THREAD

**Issue by kwojtusik** [thierrygosselin/radiator#96]
Hi Thierry,

I'm not sure if you remember but about a month ago I was having some trouble with radiator converting files to rubias (#90 ). 

I finally got back around to working with the data, and still couldn't get the genomic_converter to work, even with the correct strata file.

So I used write_rubias instead and that worked, however Rubias was having issues with the chromosome names and after some digging I found that the chromosome names were in "doubles" format instead of being "characters" as required by Rubias. So a quick conversion fixed it and then things ran smoothly. 

I'm not sure if this issue was something that was only happening with my data or if it was a bug in the package. I'd be happy to provide my files if its necessary. Thank you.

-Kris

**Comment by thierrygosselin** [comment]
Hi Kris, send the files by email and show me the command you use so that I can reproduce, I'll have a look as soon as I get the files.

Best
Thierry

**Comment by eriqande** [comment]
Hey Thierry,

It is the columns of allele data in rubias that need to be either integers
or characters.  If they aren't (i.e. if they are numeric/double) then
rubias throws an error.  Simply coercing the type of each allele column in
the rubias data frame to character takes care of it.

For example, if the genetic data starts on column 5 in the tibble dtf, then

dtf[-(1:5)] <- lapply(dtf[-(1:5)], as.character)

will take care of it.

Also, that construct with lapply turns out to be WAY faster than trying to
use mutate_at().

eric

**Comment by thierrygosselin** [comment]
ok I'll check that
Thanks Eric!

**Comment by thierrygosselin** [comment]
Kris, in the genepop file provided last time this was your markers names: `1_0,2_0,3_0,4_0,5_0,`
That should work without modification
send me at least the strata with the rubias required column

My test dataset still works with `radiator::genomic_converter` and rubias. I'd like to adapt the script to your data because someone else might run into the same problem...

**Comment by thierrygosselin** [comment]
```
data <- radiator::tidy_genepop(data = "Allpops_13k.gen.gz", strata = "All_Rivers.strata.tsv")
test1 <- radiator::write_rubias(data = data, strata = NULL, filename = "test")
```

works for me
All the columns in the object `test1` are character
The file generated is the same, depends on how it's imported but these headers:
 `sample_type	repunit	collection	indiv	10000_0.A1	10000_0.A2	10001_0.A1	10001_0.A2`

Should not be a problem for rubias

**Comment by thierrygosselin** [comment]
```
test2 <- radiator::genomic_converter(
  data = "Allpops_13k.gen.gz",
  strata = "All_Rivers.strata.tsv",
  output = "rubias"
  )
```

generates this:

```
################################################################################
######################### radiator::genomic_converter ##########################
################################################################################
Execution date@time: 20200930@1300
Folder created: 02_radiator_genomic_converter_20200930@1300
Function call and arguments stored in: radiator_genomic_converter_args_20200930@1300.tsv
Filters parameters file generated: filters_parameters_20200930@1300.tsv

Importing data

Synchronizing data and strata...
    Number of strata: 9
    Number of individuals: 438
Calibrating REF/ALT alleles...
The separator specified is not valid

Writing tidy data set:
radiator_data_20200930@1300.rad

Computation time, overall: 65 sec

Preparing data for output

Data is bi-allelic
Generating rubias output
 Error: the strata file/object requires 4 columns: SAMPLE_TYPE, REPUNIT, COLLECTION, INDIVIDUALS
Run `rlang::last_error()` to see where the error occurred. 
5.
stop(fallback) 
4.
signal_abort(cnd) 
3.
rlang::abort("the strata file/object requires 4 columns: SAMPLE_TYPE, REPUNIT, COLLECTION, INDIVIDUALS") at write_rubias.R#217
2.
radiator::write_rubias(data = input, strata = strata.bk, filename = filename, 
    parallel.core = parallel.core) at genomic_converter.R#558
1.
radiator::genomic_converter(data = "Allpops_13k.gen.gz", strata = "All_Rivers.strata.tsv", 
    output = "rubias") 
Computation time, overall: 69 sec
######################### completed genomic_converter ##########################
```


- Will fix this
- it's an additional check that I'll remove. 
- In the previous example, the required columns are filed automatically as described in the function [doc](https://thierrygosselin.github.io/radiator/reference/write_rubias.html)

**Comment by kwojtusik** [comment]
Oh, I switched to working with a VCF file. Could that be the issue?

**Comment by thierrygosselin** [comment]
Most probably!

There are so many ways to write a VCF 
Probably as much as genepop file colors...

send the VCF...

**Comment by thierrygosselin** [comment]
with the latest GitHub push genepop and vcf should work fine, with and without a strata file

**Comment by thierrygosselin** [comment]
For the vcf 
```
test <- radiator::genomic_converter(
  data = "Allpops_13k_stacks_recode.vcf",
  strata = "All_Rivers.strata.tsv",
  output = "rubias"
)
```

**Comment by thierrygosselin** [comment]
I've looked at your VCF...

```
data <- radiator::read_vcf(data = "Allpops_13k_stacks_recode.vcf", strata = "All_Rivers.strata.tsv")
```

This is pretty quick, and you should get this

```
> data <- radiator::read_vcf(data = "Allpops_13k_stacks_recode.vcf", strata = "All_Rivers.strata.tsv")
Execution date@time: 20200930@1426
Folder created: read_vcf_20200930@1426
Function call and arguments stored in: radiator_read_vcf_args_20200930@1426.tsv
File written: random.seed (214342)

Reading VCF
Data summary: 
  number of samples: 438
number of markers: 13361
done! timing: 3 sec


GDS file written: radiator_20200930@1426.gds

Analyzing the vcf...
VCF source: Stacks v2.4
Data is bi-allelic
Cleaning VCF's sample names
File written: cleaned.vcf.id.info_20200930@1426.tsv
Synchronizing data and strata...
    Number of strata: 9
    Number of individuals: 438
Reads assembly: reference-assisted
Filters parameters file generated: filters_parameters_20200930@1426.tsv
Filters parameters file: initiated and updated

Detected 182 duplicate SNPs on different strands (+/-)
    By default radiator prune those SNPs
    To change this behavior, use argument: filter.strands
File written: blacklist.duplicated.markers.strands_20200930@1426.tsv
File written: whitelist.duplicated.markers.strands_20200930@1426.tsv
Filters parameters file: updated
################################### RESULTS ####################################

Filter duplicated markers on different strands: blacklist
Number of individuals / strata / chrom / locus / SNP:
    Before: 438 / 9 / 37 / 13361 / 13361
    Blacklisted: 0 / 0 / 0 / 182 / 182
    After: 438 / 9 / 37 / 13179 / 13179
################################################################################
######################### radiator::filter_monomorphic #########################
################################################################################
Execution date@time: 20200930@1426
Function call and arguments stored in: radiator_filter_monomorphic_args_20200930@1426.tsv
Filters parameters file: initiated
File written: blacklist.monomorphic.markers_20200930@1426.tsv
Synchronizing markers.meta
File written: whitelist.polymorphic.markers_20200930@1426.tsv
Filters parameters file: updated
################################### RESULTS ####################################

Filter monomorphic markers
Number of individuals / strata / chrom / locus / SNP:
    Before: 438 / 9 / 37 / 13179 / 13179
    Blacklisted: 0 / 0 / 0 / 1 / 1
    After: 438 / 9 / 37 / 13178 / 13178

Computation time, overall: 0 sec
######################### completed filter_monomorphic #########################
################################################################################
####################### radiator::filter_common_markers ########################
################################################################################
Execution date@time: 20200930@1426
Function call and arguments stored in: radiator_filter_common_markers_args_20200930@1426.tsv
Filters parameters file: initiated
Scanning for common markers...
Generating UpSet plot to visualize markers in common
File written: blacklist.not.common.markers_20200930@1426.tsv
File written: whitelist.common.markers_20200930@1426.tsv
Filters parameters file: updated
################################### RESULTS ####################################

Filter common markers:
Number of individuals / strata / chrom / locus / SNP:
    Before: 438 / 9 / 37 / 13178 / 13178
    Blacklisted: 0 / 0 / 11 / 11460 / 11460
    After: 438 / 9 / 26 / 1718 / 1718

Computation time, overall: 2 sec
####################### completed filter_common_markers ########################

Preparing output files...
File written: whitelist.markers.tsv
File written: blacklist.markers.tsv
File written: strata.filtered.tsv
Generating individual stats...
File written: individuals qc info and stats summary
File written: individuals qc plot
Generating markers stats...
################################### SUMMARY ####################################

VCF summary):
Missing data: 
    markers: 0.1
    individuals: 0.1


Coverage info:
    individuals mean read depth: 65202
    individuals mean genotype coverage: 39
    markers mean coverage: 41


Number of chromosome/contig/scaffold: 26
Number of locus: 1718
Number of markers: 1718
Number of populations: 9
Number of individuals: 438

Number of ind/pop:
Roanoke = 20
Potomac = 35
Choptank = 43
Upper_Bay = 81
Delaware = 39
Hudson = 88
Kennebec = 42
Shubenacadie = 59
Miramichi = 31

Number of duplicate id: 0
radiator Genomic Data Structure (GDS) file: radiator_20200930@1426.gds

Computation time, overall: 16 sec
############################## completed read_vcf ##############################
```

With the default during VCF reading you end up with similar number of markers.

182 are blacklisted because they are duplicated and on different strands
1 is blacklisted because it's monomorphic
11460 are blacklisted because they are not common between the strata

**Comment by thierrygosselin** [comment]
I usually run these 2 checks...

# Check individual heterozygosity
```
het.check <- radiator::detect_mixed_genomes(data)
# this will ask if you want to remove samples I answered no
```

# Check for duplicates or technical replicates...
```
dup.check <- radiator::detect_duplicate_genomes(data)
# this will ask if you want to remove samples I answered no
```

You can run those, it's very quick. But I'll also send the folder to you.

You want to make sure before running *rubias* that you're using a filtered dataset and that a couple of metrics are kept under control...underwise you're results will be highly biased. And currently, with this dataset, I wouldn't trust the results.

**With those 2 checks I have to ask:**

1. what's the specie again? looks almost clonal, samples are very very similar...
2. wet lab related question:
- the sampling, DNA extraction, library prep and sequencing... anything odd ? 
- was the tissus of good quality ?
- how many person did the wet lab part ?
- same sequencing strategy for all samples ?


Lots of questions, but it helps me make correlation between what I read from the figures and what happened...
When I see individual heterozygosity figure like that one:

- It's usually with unfiltered dataset
- with historical DNA or degraded DNA
- extraction done by 1 or more persons with varying level of experience
- FYI I've done all of the mistakes you could think of in a wet lab...
- I also see those with sequencing done with 2 different types of machine (e.g. illumina 2500 and 4000)


**Bottom line:**
- you have a wide range of individual with very few missing data (min = 0.001) to almost ALL missing (max = 0.98), but still this produce ok average metrics, the median and IQR are low. FYI Those stats are produced automatically when the VCF is read in.
- the missingness and heterozygosity are directly linked to the sequencing effort (coverage) of your samples. You have samples with 100x times more coverage. 
- You should definitely filter the data first based on samples stats after on markers stats.
- check this paper: https://ecoevorxiv.org/47tka/
- try `??radiator::filter_rad` with the stacks VCF (put very low thresholds in stacks populations module and do the filtering in *radiator*

**Comment by kwojtusik** [comment]
Hi Thierry,

Thank you for looking at this, I really appreciate it. 

In answer to your questions (in case you were curious):

1) Species: striped bass- an anadromous fish on the east of the US.
2) wet lab: There was nothing odd about the wet lab prep. All the DNA came from a collaborator that extracted them 10-30 years ago (depending on the group). I'm not sure if one person did them all or not. I'm assuming yes, but can't be sure. I did all the library prep for these samples and they were sequenced on the same machine at the same time. 

This particular dataset I sent was one of a couple that I''m currently working with, and is not filtered considerably well (that was the idea). Looking back I probably should have sent our more stringently filtered data set, but hind sight is 20/20 i guess!!

I know there are samples, mostly from one population in particular, that have a lot of missingness and they may be thrown out entirely, we are currently working on re-sequencing to get better coverage but do not know if/when we will be able to get it done.

I do have one question, if you don't mind. What would you expect to see in the individual heterozygosity figure for a normal or good dataset? And why would you expect that? Sorry if this is a foolish question.

Thanks again for your help.

-Kris

**Comment by thierrygosselin** [comment]
Hey it's a really legitimate question, and lately it's been asked a lot...

But first, the pattern I see and the answer you've given, for me it's enough. It's the DNA quality. 10-30 years ago... it's a long time. As I mentioned before, I've seen these patterns with degraded and historical DNA. The DNA would probably be good enough for microsatellites analysis, however, RADseq studies require top quality. 

Since you've made the libraries...

- How would you rate your wet lab skills from 1 to 10 (highest)? 

Because I've also seen this... Laura who did the lobster paper that Eric Anderson like to use ;)

- She did the libraries herself. 
- When she was a master student with very little experience. 
- Never went to another lab to learn the technique. 
- Just by herself. 
- The DNA quantification protocol used at that time in the lab was also wrong. 
- Using nanodrop is not good enough and pico green is not easy...

More recently, I've seen similar patterns with my own datasets :

- Sturgeons: DNA from the same fish exhibited big differences of heterozygosity when it came from scales or tissue.

- Black Spruce: huge differences if it's coming from needles or cambium. For that last one, we also saw big differences with the same sample, but tissue taken at different ages... 

The problem with outlier heterozygote samples...

- They generate **artifact close kin** when the heterozygosity is much higher than the rest (this is where the bubble size and dotted line helps). This is clearly you're base. From the look of the duplicate plot (IBS, pairwise distances), I could have swore you were working with a mussel, oyster or stickleback fish ;)
- They generate **false population signals**: when the het is lower and correlated with missing data and coverage.

The problem of identity-by-missingness is an entirely different problem. I'm now scared more by heterozygosity pattens than by missing data patterns. For IBM analysis, have a look at my other package: [grur](https://thierrygosselin.github.io/grur/). 


My advice:

- you really shouldn't conduct downstream analysis to develop a panel of SNPs with unfiltered data. It's worst for close kin, assignment and linkage map.
- Don't get the wrong ideas, it's totally fine to test and see how the filters impact analysis :) 
- I think that the [assumptions in assigner](http://thierrygosselin.github.io/assigner/reference/assignment_ngs.html#assumptions) could also apply to rubias. Eric might have something to add on this...
- unless you want to become a lab technician, you should **stay away from the wet lab** ;) and focus on analysis, life history of your species. I know it's difficult to digest for PI and Professors that rely on cheap students...
- have very very few people work on the wet lab part
- Ideally, have robots do the steps.
- Outsource the wet lab to good companies (e.g. with DArTseq I got a sturgeon project, where the cost of DNA extraction, libraires, sequencing and genotyping (no stacks) was actually a lot cheaper than sending the DNA to IBIS Université Laval and illumina sequencing at McGill or ion torrent sequencing at IBIS.
- Please don't cook up a story in your head about the data, be critical. We don't really need another paper that tell a story that is not.

If you do more sequencing for this project:

- put the sequencing lanes or chips in [fastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) and don't hesitate to ask for re-sequencing if the company screwed up the sequencing. During a postdoc, I got 21 lanes of Illumina sequencing for a sturgeon project, out of those I asked to re-sequence 7 lanes...
- use replicates, use replicates, use replicates (e.g. the same sample twice on the same plate and the same sample on other plate if you have more than one plate. Use specific positions for the replicates to also make sure that nobody inverse the plate...
- to increase coverage: don't pool the same samples from the same/different lanes/chips, it makes everything worst.
- instead, reduce the number of samples per library/lanes.

Later today, I'll send a talk I gave at NOAA about heterozygosity and in there you'll see good and bad examples of individual heterozygosity. You'll also see the impact of having a mix of two species in those plots (I've got a squid and a tuna example). Like I said earlier, I've done ALL the mistakes...

Hope this helps...

**Comment by thierrygosselin** [comment]
I've sent the keynote, let me know if it make sense or not.
This week I had to deal with a sturgeon dataset of mine.
The sampling was conducted in a way that out of the 7 sampling sites you have:

-  1 site with dry fin
- 4 sites with fin preserved in ETOH
- 2 sites preserved with RNA-later
- 2 sites with 10 years of sampling (this impact DNA quality...)
- I inherited the project after sequencing was done...
- The samples were not randomized for library prep or sequencing
- Instead of doing it right the first time the researchers involved for that part decided to sequence on Ion Torrent using 4 to 7 chips. 
- So I have a bunch of samples sequence 4 times and another batch 7 times.
- Despite the sequencing effort and having 3M reads/samples I have 10-15 x coverage / samples at the end.
- It's just ridiculous... and people are disappointed because I'm telling them that close kin analysis is out of the question...
